core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -727,12 +732,25 @@ private Type handleResolvedField(Expression node, ResolvedField resolvedField, S
 
             FieldId fieldId = FieldId.from(resolvedField);
             Field field = resolvedField.getField();
+            Type fieldType = fieldId.getRelationId()
+                    .getSourceNode()
+                    .map(sourceNode -> {
+                        if (sourceNode instanceof Relation relation) {
+                            List<Type> coercion = getRelationCoercion.apply(relation);
+                            if (coercion != null) {
+                                return coercion.get(resolvedField.getRelationFieldIndex());
+                            }
+                        }
+                        return field.getType();
+                    })
+                    .orElseGet(field::getType);
+
```

Comment:
Formatting nit:
```suggestion
            Type fieldType = field.getType();
            Optional<Node> sourceNode = fieldId.getRelationId().getSourceNode();
            if (sourceNode.isPresent() && sourceNode.get() instanceof Relation relation) {
                List<Type> coercions = getRelationCoercion.apply(relation);
                if (coercions != null) {
                    fieldType = coercions.get(resolvedField.getRelationFieldIndex());
                }
            }
```

Reply:
Good call, and that way i can use sourceNode further down the method.